image/image_linux.go

```diff
@@ -692,29 +613,196 @@ var setupSeed = func(tsto *tooling.ToolingStore, model *asserts.Model, opts *Opt
 		return err
 	}
 
-	if err := boot.MakeBootableImage(model, bootRootDir, bootWith, opts.Customizations.BootFlags); err != nil {
+	if err := boot.MakeBootableImage(is.model, is.bootRootDir, bootWith, opts.Customizations.BootFlags); err != nil {
 		return err
 	}
 
 	// early config & cloud-init config (done at install for Core 20)
-	if !hasModes {
+	if !is.hasModes {
 		// and the cloud-init things
-		if err := installCloudConfig(rootDir, gadgetUnpackDir); err != nil {
+		if err := installCloudConfig(is.rootDir, gadgetUnpackDir); err != nil {
 			return err
 		}
 
-		defaultsDir := sysconfig.WritableDefaultsDir(rootDir)
+		defaultsDir := sysconfig.WritableDefaultsDir(is.rootDir)
 		defaults := gadget.SystemDefaults(gadgetInfo.Defaults)
 		if len(defaults) > 0 {
-			if err := os.MkdirAll(sysconfig.WritableDefaultsDir(rootDir, "/etc"), 0755); err != nil {
+			if err := os.MkdirAll(sysconfig.WritableDefaultsDir(is.rootDir, "/etc"), 0755); err != nil {
 				return err
 			}
-			if err := sysconfig.ApplyFilesystemOnlyDefaults(model, defaultsDir, defaults); err != nil {
+			if err := sysconfig.ApplyFilesystemOnlyDefaults(is.model, defaultsDir, defaults); err != nil {
 				return err
 			}
 		}
 
-		customizeImage(rootDir, defaultsDir, &opts.Customizations)
+		customizeImage(is.rootDir, defaultsDir, &opts.Customizations)
+	}
+	return nil
+}
+
+func optionSnaps(opts *Options) []*seedwriter.OptionsSnap {
+	optSnaps := make([]*seedwriter.OptionsSnap, 0, len(opts.Snaps))
+	for _, snapName := range opts.Snaps {
+		var optSnap seedwriter.OptionsSnap
+		if strings.HasSuffix(snapName, ".snap") {
+			// local
+			optSnap.Path = snapName
+		} else {
+			optSnap.Name = snapName
+		}
+		optSnap.Channel = opts.SnapChannels[snapName]
+		optSnaps = append(optSnaps, &optSnap)
+	}
+	return optSnaps
+}
+
+var setupSeed = func(tsto *tooling.ToolingStore, model *asserts.Model, opts *Options) error {
+	if model.Classic() != opts.Classic {
+		return fmt.Errorf("internal error: classic model but classic mode not set")
+	}
+
+	is, err := makeImageSeeder(tsto, model, opts)
+	if err != nil {
+		return err
+	}
+
+	optSnaps := optionSnaps(opts)
+	if err := is.w.SetOptionsSnaps(optSnaps); err != nil {
+		return err
+	}
+
+	// TODO: developer database in home or use snapd (but need
+	// a bit more API there, potential issues when crossing stores/series)
+	db, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
+		Backstore: asserts.NewMemoryBackstore(),
+		Trusted:   trusted,
+	})
+	if err != nil {
+		return err
+	}
+
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		return tsto.AssertionFetcher(db, save)
+	}
+	f, err := is.w.Start(db, newFetcher)
```

Comment:
potentially this could be organized into a imageSeeder.Start method that takes a db and optSnaps?

Reply:
Nice suggestion, went ahead and introduced .start and .finsh methods